File: portscanner.py
#!/bin/python3

import socket
import termcolor
import ipaddress
import pyfiglet

def scan_ports(target, start_port, end_port):
    open_ports = []
    for port in range(start_port, end_port + 1):
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.settimeout(1)
        result = sock.connect_ex((target, port))
        if result == 0:
            open_ports.append(port)
        sock.close()
    return open_ports

def main():
    ascii_banner = pyfiglet.figlet_format("tony's port scanner")
    print(ascii_banner)
    
    targets = input("Enter the target(s) to scan: ")
    target_list = targets.split(',')
    
    start_port = int(input("Enter the starting port: "))
    end_port = int(input("Enter the ending port: "))
    
    for target in target_list:
        target = target.strip()
        print(f"\nScanning ports for {target}...")
        
        if '/' in target:  # CIDR range
            ip_range = ipaddress.IPv4Network(target, strict=False)
            for ip in ip_range.hosts():
                open_ports = scan_ports(str(ip), start_port, end_port)
                if open_ports:
                    print(f"\n Open ports for {ip}:")
                    for port in open_ports:
                        print(termcolor.colored(f"[*]   Port {port} is open", 'green'))
        else:  # Single host
            open_ports = scan_ports(target, start_port, end_port)
            if open_ports:
                print("\n[*] Open ports:")
                for port in open_ports:
                    print(termcolor.colored(f"[*]   Port {port} is open", 'green'))
            else:
                print(f"No open ports found for {target}.")

    scan_another = input("\nDo you want to scan another target? (yes/no): ")
    if scan_another.lower() == "yes":
        main()
    else:
        print("Exiting the port scanner.")

if __name__ == "__main__":
    main()

File: portscanner.go
package main

import (
	"bufio"
	"fmt"
	"net"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

// Function to scan a specific port and perform banner grabbing
func scanPort(ip string, port int, wg *sync.WaitGroup) {
	defer wg.Done()

	address := fmt.Sprintf("%s:%d", ip, port)
	conn, err := net.Dial("tcp", address)
	if err != nil {
		return // Port is closed
	}
	defer conn.Close()

	// Set a timeout for reading from the connection
	conn.SetReadDeadline(time.Now().Add(2 * time.Second))

	// Read the banner from the connection
	banner, err := bufio.NewReader(conn).ReadString('\n')
	if err != nil {
		return // Unable to read banner
	}

	fmt.Printf("[+] Port %d is open on %s - %s\n", port, ip, strings.TrimSpace(banner))
}

func main() {
	fmt.Print("Enter the target host or CIDR range: ")
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	target := scanner.Text()

	// Check if the target is a CIDR range
	ips, ipNet, err := net.ParseCIDR(target)
	if err != nil {
		// If not a CIDR range, use the target as is
		ip := net.ParseIP(target)
		if ip == nil {
			fmt.Println("Invalid target. Exiting.")
			return
		}
		target = ip.String()
	} else {
		// If CIDR range, use the network address and iterate over hosts
		target = ips.String()
	}

	fmt.Print("Enter the starting port: ")
	scanner.Scan()
	startPort, _ := strconv.Atoi(scanner.Text())

	fmt.Print("Enter the ending port: ")
	scanner.Scan()
	endPort, _ := strconv.Atoi(scanner.Text())

	// Use a WaitGroup to wait for all goroutines to finish
	var wg sync.WaitGroup

	// If the target is a CIDR range, iterate over hosts in the range
	if ipNet != nil {
		for ip := ips.Mask(ipNet.Mask); ipNet.Contains(ip); inc(ip) {
			for port := startPort; port <= endPort; port++ {
				// Increment the WaitGroup counter
				wg.Add(1)

				// Launch a goroutine for each port and IP combination
				go scanPort(ip.String(), port, &wg)
			}
		}
	} else {
		// If the target is a single host, iterate over ports
		for port := startPort; port <= endPort; port++ {
			// Increment the WaitGroup counter
			wg.Add(1)

			// Launch a goroutine for each port
			go scanPort(target, port, &wg)
		}
	}

	// Wait for all goroutines to finish
	wg.Wait()
}

// Function to increment an IP address
func inc(ip net.IP) {
	for j := len(ip) - 1; j >= 0; j-- {
		ip[j]++
		if ip[j] > 0 {
			break
		}
	}
}
